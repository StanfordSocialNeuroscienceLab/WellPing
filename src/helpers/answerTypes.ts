import { AnswerEntity } from "../entities/AnswerEntity";
import { Question, QuestionsList, Choice } from "./types";

// We use `{ value: ... }` because it seems that `"simple-json"`
// does not support simply mixing normal values.
// This also increases the consistency of the data.

export type SliderAnswerData = {
  value: number;
};

export type ChoicesWithSingleAnswerAnswerData = {
  value: Choice;
};

// We use an array of tuples here (instead of `{ [key: string]: boolean }`)
// because the order shown on the screen can be kept in the results.
// This is potentially helpful for analysis later if the order is randomized
// as now we know the order the user saw the choices in.
export type ChoicesWithMultipleAnswersAnswerChoices = [Choice, boolean][];
export type ChoicesWithMultipleAnswersAnswerData = {
  value: ChoicesWithMultipleAnswersAnswerChoices;
};

export type YesNoAnswerData = {
  value: boolean;
};

export type MultipleTextAnswerData = {
  value: string[];
};

export type HowLongAgoAnswerDataType = [number | null, string | null];
export type HowLongAgoAnswerData = {
  value: HowLongAgoAnswerDataType;
};

export type AnswerData =
  | SliderAnswerData
  | ChoicesWithSingleAnswerAnswerData
  | ChoicesWithMultipleAnswersAnswerData
  | YesNoAnswerData
  | MultipleTextAnswerData
  | HowLongAgoAnswerData;

export interface AnswersList {
  [id: string]: AnswerEntity;
}

export interface QuestionScreenProps {
  question: Question;
  onDataChange: (data: AnswerData) => void;
  allAnswers: AnswersList;
  allQuestions: QuestionsList;
  pipeInExtraMetaData: (input: string) => string;
  setDataValidationFunction: (func: () => boolean) => void;
}
