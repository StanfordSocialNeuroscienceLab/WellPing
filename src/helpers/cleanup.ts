import { secureRemoveAllAnswersAsync } from "./secureStore/answer";
import { secureRemoveAllPingsAsync } from "./secureStore/ping";

export async function clearAllPingsAndAnswersAsync(): Promise<void> {
  await secureRemoveAllAnswersAsync();
  await secureRemoveAllPingsAsync();
}
