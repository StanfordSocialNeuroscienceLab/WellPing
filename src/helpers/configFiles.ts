import { isThisWeek } from "date-fns";

import { SurveyFile, Names, StudyID, StudyInfo, StreamName } from "./types";

export function getSurveyFile(): SurveyFile {
  const survey: SurveyFile = require("../../config/survey.json");
  return survey;
}

// TODO: DECOUPLE FUNCTIONS LIKE THIS
export function getNamesFile(): Names {
  const names: Names = require("../../config/names.json");
  return names;
}

export function getStudyInfo(): StudyInfo {
  return getSurveyFile().studyInfo;
}

export function getAllStreamNames(): StreamName[] {
  return Object.keys(getSurveyFile().meta.startingQuestionIds) as StreamName[];
}

export function getStudyId(): StudyID {
  return getStudyInfo().id;
}

export function isTimeThisWeek(time: Date): boolean {
  return isThisWeek(time, { weekStartsOn: getStudyInfo().weekStartsOn });
}
