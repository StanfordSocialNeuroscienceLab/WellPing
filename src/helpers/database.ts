import * as FileSystem from "expo-file-system";
import { Share } from "react-native";
import { createConnection, Connection, getConnection } from "typeorm";

import {
  AnswerEntity,
  SliderAnswerEntity,
  ChoicesWithSingleAnswerAnswerEntity,
  ChoicesWithMultipleAnswersAnswerEntity,
  YesNoAnswerEntity,
  MultipleTextAnswerEntity,
  HowLongAgoAnswerEntity,
} from "../entities/AnswerEntity";
import { PingEntity } from "../entities/PingEntity";
import { EncryptedColumnSubscriber } from "../helpers/typeormEncryption/EncryptedColumnSubscriber";
import {
  getCriticalProblemTextForUser,
  alertWithShareButtonContainingDebugInfo,
} from "./debug";

// TODO: remove me once typeorm is updated to a newer version (https://github.com/typeorm/typeorm/issues/6233)

const getDatabaseFilename = (databaseName: string) => `${databaseName}.db`;

export const entities = [
  AnswerEntity,
  SliderAnswerEntity,
  ChoicesWithSingleAnswerAnswerEntity,
  ChoicesWithMultipleAnswersAnswerEntity,
  YesNoAnswerEntity,
  MultipleTextAnswerEntity,
  HowLongAgoAnswerEntity,
  PingEntity,
];

export async function connectDatabaseAsync(
  databaseName: string,
): Promise<Connection> {
  try {
    return getConnection();
  } catch (e) {
    try {
      return await createConnection({
        type: "expo",
        driver: require("expo-sqlite"),
        database: getDatabaseFilename(databaseName),
        entities,
        subscribers: [EncryptedColumnSubscriber],
        synchronize: true,
        logging: __DEV__, // Only log in dev mode.
      });
    } catch (e) {
      alertWithShareButtonContainingDebugInfo(
        getCriticalProblemTextForUser(`connectDatabaseAsync: ${e}`),
      );
      throw e;
    }
  }
}

export function getDatabaseFolderUrl(): string {
  // https://forums.expo.io/t/how-to-upload-sqlite-db-dump/2315/3
  return `${FileSystem.documentDirectory}/SQLite/`;
}
export function getDatabaseFileUrl(databaseName: string): string {
  // https://forums.expo.io/t/how-to-upload-sqlite-db-dump/2315/3
  return `${getDatabaseFolderUrl()}${getDatabaseFilename(databaseName)}`;
}

export async function databaseFileExistsAsync(
  databaseName: string,
): Promise<boolean> {
  return (await FileSystem.getInfoAsync(getDatabaseFileUrl(databaseName)))
    .exists;
}

export async function shareDatabaseFileAsync(databaseName: string) {
  if (!(await databaseFileExistsAsync(databaseName))) {
    alert(`Database "${databaseName}" does not exists!`);
  }

  // https://forums.expo.io/t/how-to-upload-sqlite-db-dump/2315/3
  Share.share({
    url: getDatabaseFileUrl(databaseName),
  });
}

export async function deleteDatabaseFileAsync(databaseName: string) {
  if (!(await databaseFileExistsAsync(databaseName))) {
    return;
  }

  await FileSystem.deleteAsync(getDatabaseFileUrl(databaseName));
}

export async function backupDatabaseFileAsync(databaseName: string) {
  if (!(await databaseFileExistsAsync(databaseName))) {
    // Don't do anything if the database file does not exists.
    return;
  }

  await FileSystem.copyAsync({
    from: getDatabaseFileUrl(databaseName),
    to: getDatabaseFileUrl(`${databaseName}.${new Date().getTime()}.bk`),
  });
}

export async function getDatabaseFolderFilelistAsync() {
  return await FileSystem.readDirectoryAsync(getDatabaseFolderUrl());
}
