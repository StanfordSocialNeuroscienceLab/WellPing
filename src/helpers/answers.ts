import { AnswerData, Answer } from "./answerTypes";
import { getAnswersPingIdsQuestionIdsListAsync } from "./asyncStorage/answersPingIdsQuestionIdsList";
import { AnswerSchema } from "./schemas/Answer";
import {
  secureGetAnswerAsync,
  secureStoreAnswerAsync,
} from "./secureStore/answer";
import { Question, Ping } from "./types";

export async function getAnswersAsync(): Promise<Answer[]> {
  const answersList = await getAnswersPingIdsQuestionIdsListAsync();

  // https://stackoverflow.com/q/28066429/2603230
  const answers: Answer[] = await Promise.all(
    answersList.map(async (pingIdAndQuestionId) => {
      const answer = await secureGetAnswerAsync(
        pingIdAndQuestionId[0],
        pingIdAndQuestionId[1],
      );
      if (answer === null) {
        throw new Error(
          "answer === null in answersList.map in getAnswersAsync.",
        );
      }
      return answer;
    }),
  );

  return answers;
}

export async function insertAnswerAsync({
  ping,
  question,
  realQuestionId,
  preferNotToAnswer,
  data,
  date,
}: {
  ping: Ping;
  question: Question;
  realQuestionId: string;
  preferNotToAnswer: true | null; // See `MARK: WHY_PNA_TRUE_OR_NULL`.
  data: AnswerData | null;
  date: Date;
}): Promise<Answer> {
  const answer = AnswerSchema.parse({
    pingId: ping.id,
    questionId: realQuestionId,
    preferNotToAnswer,
    data,
    date,
  });

  await secureStoreAnswerAsync(answer);

  return answer;
}
