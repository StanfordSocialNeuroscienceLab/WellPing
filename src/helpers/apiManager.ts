import Constants from "expo-constants";
import * as Crypto from "expo-crypto";
import * as Device from "expo-device";
import { Platform } from "react-native";

import { AnswerEntity } from "../entities/AnswerEntity";
import { PingEntity } from "../entities/PingEntity";
import { getPingsAsync, getAnswersAsync } from "./asyncStorage";
import { getSurveyFileAsync } from "./configFiles";
import { getUserAsync, User, storeUserAsync } from "./user";

export async function getServerUrlAsync(): Promise<string> {
  return (await getSurveyFileAsync()).studyInfo.serverURL;
}

type UploadData = {
  patientId: string;
  pings: PingEntity[];
  answers: AnswerEntity[];
};

// If success, return `null`. Else return error message.
export async function registerUserAsync(user: User): Promise<string | null> {
  if (!user.patientId || !user.password) {
    return "You should enter both User ID and password.";
  }

  const request: { [key: string]: any } = {
    new_password: user.password, // Do not reset password
    phone_number: "1234567890",
    device_os: Device.osName,
    os_version: Device.osVersion,
    product: Device.modelName,
    brand: Device.brand,
    manufacturer: Device.manufacturer,
    model: Device.modelId,
    beiwe_version: "1.6.0",
  };

  try {
    let endpoint = "/register_user";
    if (Platform.OS === "ios") {
      endpoint += "/ios/";
    }
    await makePostRequestAsync(endpoint, request, {}, user);
  } catch (e) {
    return `Request error: ${e}.`;
  }

  await storeUserAsync(user);
  return null;
}

export async function getAllDataAsync(): Promise<UploadData> {
  const user = await getUserAsync();
  const pings = await getPingsAsync();
  const answers = await getAnswersAsync();

  const data = {
    patientId: user!.patientId,
    pings,
    answers,
  };
  return data;
}

export async function uploadDataAsync() {
  const user = (await getUserAsync())!;
  const data = await getAllDataAsync();

  try {
    let endpoint = "/ssnl_upload";
    if (Platform.OS === "ios") {
      endpoint += "/ios/";
    }
    const response = await makePostRequestAsync(endpoint, {}, data, user);
    //console.warn(`YAYA ${JSON.stringify(response)}`);
    return { status: "success", response };
  } catch (e) {
    return { status: "error", error: `Request error: ${e}.` };
  }
}

function base64ToBase64URL(input: string): string {
  return input.replace(/\+/g, "-").replace(/\//g, "_");
}

export async function getRequestURLAsync(
  endpoint: string,
  request: { [key: string]: any } = {},
  forUser?: User,
): Promise<string> {
  let user: User | null = forUser || null;
  if (!user) {
    user = await getUserAsync();

    if (user == null) {
      throw new Error("user == null in getRequestURLAsync");
    }
  }

  const passwordHash = await Crypto.digestStringAsync(
    Crypto.CryptoDigestAlgorithm.SHA256,
    user.password,
    {
      encoding: Crypto.CryptoEncoding.BASE64,
    },
  );
  request["password"] = base64ToBase64URL(passwordHash);
  request["device_id"] = `${user.patientId}-${Constants.installationId}`;
  request["patient_id"] = user.patientId;

  //console.warn(`request is ${JSON.stringify(request)}`);

  const serverUrl = await getServerUrlAsync();
  const query = Object.keys(request)
    .map((k) => encodeURIComponent(k) + "=" + encodeURIComponent(request[k]))
    .join("&");
  const url = `${serverUrl}${endpoint}?${query}`;
  return url;
}

export async function makePostRequestAsync(
  endpoint: string,
  request: { [key: string]: any },
  body?: { [key: string]: any },
  user?: User,
): Promise<any> {
  const headers = {
    "Beiwe-Api-Version": "2",
    Accept: "application/vnd.beiwe.api.v2, application/json",
  };

  const url = await getRequestURLAsync(endpoint, request, user);

  //console.warn("url is " + url);

  const response = await fetch(url, {
    method: "POST",
    headers,
    body: JSON.stringify(body),
  });

  if (response.status < 200 || response.status >= 400) {
    console.warn(`ERRORREQUEST: ${JSON.stringify(response)}`);

    if (response.status === 401 || response.status === 403) {
      throw new Error(
        `Verification failed.\n\nRequest: ${JSON.stringify(
          request,
        )}.\n\nResponse: ${JSON.stringify(response)}.`,
      );
    }

    throw new Error(
      `Request not successful.\n\nRequest: ${JSON.stringify(
        request,
      )}.\n\nResponse: ${JSON.stringify(response)}.`,
    );
  }

  //console.warn(response);

  return response.json();
}
