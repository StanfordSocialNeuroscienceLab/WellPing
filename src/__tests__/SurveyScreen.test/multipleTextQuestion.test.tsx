/**
 * Notice that UI elements are not tested here as they are tested in
 * MultipleTextQuestionScreen.test.tsx.
 */
import React from "react";
import { render, RenderAPI } from "react-native-testing-library";

import SurveyScreen, { SurveyScreenProps } from "../../SurveyScreen";
import { QuestionType } from "../../helpers/helpers";
import { changeTextAndWaitForUpdateAsync } from "../reactNativeTestingLibraryHelper";
import {
  TEST_PING,
  mockNecessaryFunctionsToTestSurveyScreen,
  testQuestionsSequenceAsync,
} from "./helper";

beforeEach(() => {
  mockNecessaryFunctionsToTestSurveyScreen();
});

async function enterTextAsync(
  textFieldIndex: number,
  text: string,
  { findByA11yLabel }: RenderAPI,
) {
  await changeTextAndWaitForUpdateAsync(
    // `findBy` does the `waitFor` for us.
    async () => await findByA11yLabel(`text input ${textFieldIndex}`),
    text,
  );
}

const propsBase: SurveyScreenProps = {
  questions: {},
  startingQuestionId: "q1",
  ping: TEST_PING,
  previousState: null,
  onFinish: async () => {},
};

describe("without repeatedItemStartId", () => {
  const props: SurveyScreenProps = {
    ...propsBase,
    questions: {
      q1: {
        id: "q1",
        type: QuestionType.MultipleText,
        question: "Question 1",
        indexName: "INDEX",
        variableName: "VARNAME",
        max: 3,
        next: "q2",
      },
      q2: {
        id: "q2",
        type: QuestionType.Slider,
        question: "Question 2",
        slider: ["left", "right"],
        next: null,
      },
    },
  };

  test("enter some text", async () => {
    const onFinishFn = jest.fn();

    const renderResults = render(
      <SurveyScreen {...props} onFinish={onFinishFn} />,
    );

    await testQuestionsSequenceAsync({
      renderResults,
      onFinishFn,
      sequence: [
        {
          expectCurrentQuestionAsync: async (getCurrentQuestionTitle) => {
            await enterTextAsync(0, "hello world", renderResults);
            await enterTextAsync(1, "yes?", renderResults);

            expect(getCurrentQuestionTitle()).toBe("Question 1");
          },
        },
        {
          expectCurrentQuestionAsync: async (getCurrentQuestionTitle) => {
            expect(getCurrentQuestionTitle()).toBe("Question 2");
          },
        },
      ],
    });
  });

  test("enter some text and then remove them", async () => {
    const onFinishFn = jest.fn();

    const renderResults = render(
      <SurveyScreen {...props} onFinish={onFinishFn} />,
    );

    await testQuestionsSequenceAsync({
      renderResults,
      onFinishFn,
      sequence: [
        {
          expectCurrentQuestionAsync: async (getCurrentQuestionTitle) => {
            await enterTextAsync(0, "hello world", renderResults);
            await enterTextAsync(1, "yes?", renderResults);
            await enterTextAsync(0, "", renderResults);
            await enterTextAsync(1, "", renderResults);

            expect(getCurrentQuestionTitle()).toBe("Question 1");
          },
        },
        {
          expectCurrentQuestionAsync: async (getCurrentQuestionTitle) => {
            expect(getCurrentQuestionTitle()).toBe("Question 2");
          },
        },
      ],
    });
  });

  test("click next without answering", async () => {
    const onFinishFn = jest.fn();

    const renderResults = render(
      <SurveyScreen {...props} onFinish={onFinishFn} />,
    );

    await testQuestionsSequenceAsync({
      renderResults,
      onFinishFn,
      sequence: [
        {
          expectCurrentQuestionAsync: async (getCurrentQuestionTitle) => {
            expect(getCurrentQuestionTitle()).toBe("Question 1");
          },
        },
        {
          expectCurrentQuestionAsync: async (getCurrentQuestionTitle) => {
            expect(getCurrentQuestionTitle()).toBe("Question 2");
          },
        },
      ],
    });
  });

  test("click prefer not to answer", async () => {
    const onFinishFn = jest.fn();

    const renderResults = render(
      <SurveyScreen {...props} onFinish={onFinishFn} />,
    );

    await testQuestionsSequenceAsync({
      renderResults,
      onFinishFn,
      sequence: [
        {
          expectCurrentQuestionAsync: async (getCurrentQuestionTitle) => {
            // Entering a text shouldn't matter here if we click prefer not
            // to answer.
            await enterTextAsync(0, "hello!", renderResults);

            expect(getCurrentQuestionTitle()).toBe("Question 1");
          },
          nextButton: "pna",
        },
        {
          expectCurrentQuestionAsync: async (getCurrentQuestionTitle) => {
            expect(getCurrentQuestionTitle()).toBe("Question 2");
          },
        },
      ],
    });
  });
});

describe("with repeatedItemStartId", () => {
  const props: SurveyScreenProps = {
    ...propsBase,
    questions: {
      q1: {
        id: "q1",
        type: QuestionType.MultipleText,
        question: "Question 1",
        indexName: "INDEX",
        variableName: "MY_VARNAME",
        repeatedItemStartId: "q1_n[__INDEX__]_1",
        max: 3,
        next: "q2",
      },
      "q1_n[__INDEX__]_1": {
        id: "q1_n[__INDEX__]_1",
        type: QuestionType.Slider,
        question: "Question 1-1: [__MY_VARNAME__] (index [__INDEX__])",
        slider: ["left", "right"],
        next: "q1_n[__INDEX__]_2",
      },
      "q1_n[__INDEX__]_2": {
        id: "q1_n[__INDEX__]_2",
        type: QuestionType.Slider,
        question: "Question 1-2: [__MY_VARNAME__] (index is [__INDEX__])",
        slider: ["left", "right"],
        next: null,
      },
      q2: {
        id: "q2",
        type: QuestionType.Slider,
        question: "Question 2",
        slider: ["left", "right"],
        next: null,
      },
    },
  };

  test("enter some text", async () => {
    const onFinishFn = jest.fn();

    const renderResults = render(
      <SurveyScreen {...props} onFinish={onFinishFn} />,
    );

    await testQuestionsSequenceAsync({
      renderResults,
      onFinishFn,
      sequence: [
        {
          expectCurrentQuestionAsync: async (getCurrentQuestionTitle) => {
            await enterTextAsync(0, "hello world!", renderResults);
            await enterTextAsync(1, "yes?", renderResults);

            expect(getCurrentQuestionTitle()).toBe("Question 1");
          },
        },
        {
          expectCurrentQuestionAsync: async (getCurrentQuestionTitle) => {
            expect(getCurrentQuestionTitle()).toBe(
              "Question 1-1: hello world! (index 1)",
            );
          },
        },
        {
          expectCurrentQuestionAsync: async (getCurrentQuestionTitle) => {
            expect(getCurrentQuestionTitle()).toBe(
              "Question 1-2: hello world! (index is 1)",
            );
          },
        },
        {
          expectCurrentQuestionAsync: async (getCurrentQuestionTitle) => {
            expect(getCurrentQuestionTitle()).toBe(
              "Question 1-1: yes? (index 2)",
            );
          },
        },
        {
          expectCurrentQuestionAsync: async (getCurrentQuestionTitle) => {
            expect(getCurrentQuestionTitle()).toBe(
              "Question 1-2: yes? (index is 2)",
            );
          },
        },
        {
          expectCurrentQuestionAsync: async (getCurrentQuestionTitle) => {
            expect(getCurrentQuestionTitle()).toBe("Question 2");
          },
        },
      ],
    });
  });

  test("enter some text and then remove them", async () => {
    const onFinishFn = jest.fn();

    const renderResults = render(
      <SurveyScreen {...props} onFinish={onFinishFn} />,
    );

    await testQuestionsSequenceAsync({
      renderResults,
      onFinishFn,
      sequence: [
        {
          expectCurrentQuestionAsync: async (getCurrentQuestionTitle) => {
            await enterTextAsync(0, "hello world", renderResults);
            await enterTextAsync(1, "yes?", renderResults);
            await enterTextAsync(0, "", renderResults);
            await enterTextAsync(1, "", renderResults);

            expect(getCurrentQuestionTitle()).toBe("Question 1");
          },
        },
        {
          expectCurrentQuestionAsync: async (getCurrentQuestionTitle) => {
            expect(getCurrentQuestionTitle()).toBe("Question 2");
          },
        },
      ],
    });
  });

  test("enter some text and then remove one of them", async () => {
    const onFinishFn = jest.fn();

    const renderResults = render(
      <SurveyScreen {...props} onFinish={onFinishFn} />,
    );

    await testQuestionsSequenceAsync({
      renderResults,
      onFinishFn,
      sequence: [
        {
          expectCurrentQuestionAsync: async (getCurrentQuestionTitle) => {
            await enterTextAsync(0, "後之視今", renderResults);
            await enterTextAsync(1, "亦猶今之視昔", renderResults);
            await enterTextAsync(2, "悲夫", renderResults);
            await enterTextAsync(0, "", renderResults);
            await enterTextAsync(1, "", renderResults);

            expect(getCurrentQuestionTitle()).toBe("Question 1");
          },
        },
        {
          expectCurrentQuestionAsync: async (getCurrentQuestionTitle) => {
            expect(getCurrentQuestionTitle()).toBe(
              "Question 1-1: 悲夫 (index 1)",
            );
          },
        },
        {
          expectCurrentQuestionAsync: async (getCurrentQuestionTitle) => {
            expect(getCurrentQuestionTitle()).toBe(
              "Question 1-2: 悲夫 (index is 1)",
            );
          },
        },
        {
          expectCurrentQuestionAsync: async (getCurrentQuestionTitle) => {
            expect(getCurrentQuestionTitle()).toBe("Question 2");
          },
        },
      ],
    });
  });

  test("click next without answering", async () => {
    const onFinishFn = jest.fn();

    const renderResults = render(
      <SurveyScreen {...props} onFinish={onFinishFn} />,
    );

    await testQuestionsSequenceAsync({
      renderResults,
      onFinishFn,
      sequence: [
        {
          expectCurrentQuestionAsync: async (getCurrentQuestionTitle) => {
            expect(getCurrentQuestionTitle()).toBe("Question 1");
          },
        },
        {
          expectCurrentQuestionAsync: async (getCurrentQuestionTitle) => {
            expect(getCurrentQuestionTitle()).toBe("Question 2");
          },
        },
      ],
    });
  });

  test("click prefer not to answer", async () => {
    const onFinishFn = jest.fn();

    const renderResults = render(
      <SurveyScreen {...props} onFinish={onFinishFn} />,
    );

    await testQuestionsSequenceAsync({
      renderResults,
      onFinishFn,
      sequence: [
        {
          expectCurrentQuestionAsync: async (getCurrentQuestionTitle) => {
            // Entering a text shouldn't matter here if we click prefer not
            // to answer.
            await enterTextAsync(0, "hello!", renderResults);

            expect(getCurrentQuestionTitle()).toBe("Question 1");
          },
          nextButton: "pna",
        },
        {
          expectCurrentQuestionAsync: async (getCurrentQuestionTitle) => {
            expect(getCurrentQuestionTitle()).toBe("Question 2");
          },
        },
      ],
    });
  });
});
