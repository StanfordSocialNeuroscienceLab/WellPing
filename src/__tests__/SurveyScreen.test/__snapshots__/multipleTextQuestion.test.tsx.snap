// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`with repeatedItemStartId click next without answering: getAnswersAsync 1`] = `
Array [
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q2",
  },
]
`;

exports[`with repeatedItemStartId click prefer not to answer: getAnswersAsync 1`] = `
Array [
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": true,
    "questionId": "q1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q2",
  },
]
`;

exports[`with repeatedItemStartId enter some text and then remove one of them: getAnswersAsync 1`] = `
Array [
  Object {
    "data": Object {
      "value": Array [
        "悲夫",
      ],
    },
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q1_n1_1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q1_n1_2",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q2",
  },
]
`;

exports[`with repeatedItemStartId enter some text and then remove them: getAnswersAsync 1`] = `
Array [
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q2",
  },
]
`;

exports[`with repeatedItemStartId enter some text: getAnswersAsync 1`] = `
Array [
  Object {
    "data": Object {
      "value": Array [
        "hello world!",
        "yes?",
      ],
    },
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q1_n1_1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q1_n1_2",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q1_n2_1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q1_n2_2",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q2",
  },
]
`;

exports[`without repeatedItemStartId click next without answering: getAnswersAsync 1`] = `
Array [
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q2",
  },
]
`;

exports[`without repeatedItemStartId click prefer not to answer: getAnswersAsync 1`] = `
Array [
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": true,
    "questionId": "q1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q2",
  },
]
`;

exports[`without repeatedItemStartId enter some text and then remove them: getAnswersAsync 1`] = `
Array [
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q2",
  },
]
`;

exports[`without repeatedItemStartId enter some text: getAnswersAsync 1`] = `
Array [
  Object {
    "data": Object {
      "value": Array [
        "hello world",
        "yes?",
      ],
    },
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q1",
  },
  Object {
    "data": null,
    "date": 1970-01-01T00:00:00.000Z,
    "pingId": "[removed for snapshot]",
    "preferNotToAnswer": null,
    "questionId": "q2",
  },
]
`;
